-Zmir-opt-level >= 3 errors with recursive async

If you try compiling something that includes a recursive async function with -Zmir-opt-level=3 or higher, rustc gives an error.

```
pub async fn foo(n: usize) {
    if n > 0 {
        Box::pin(foo(n - 1)).await;
    }
}
```

```
error[E0733]: recursion in an async fn requires boxing
 --> src/main.rs:1:1
  |
1 | pub async fn foo(n: usize) {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: a recursive `async fn` call must introduce indirection such as `Box::pin` to avoid an infinitely sized future
```

It looks like the responsible pass is DataflowConstProp where coroutines should probably be skipped like others do.

--- a/compiler/rustc_mir_transform/src/dataflow_const_prop.rs
+++ b/compiler/rustc_mir_transform/src/dataflow_const_prop.rs
@@ -41,6 +41,11 @@ impl<'tcx> crate::MirPass<'tcx> for DataflowConstProp {
 
     #[instrument(skip_all level = "debug")]
     fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        // Avoid query cycles from coroutines.
+        if body.coroutine.is_some() {
+            return;
+        }
+
         debug!(def_id = ?body.source.def_id());
         if tcx.sess.mir_opt_level() < 4 && body.basic_blocks.len() > BLOCK_LIMIT {
             debug!("aborted dataflow const prop due too many basic blocks");

--- a/tests/ui/dataflow_const_prop/recursive-async.rs
+++ b/tests/ui/dataflow_const_prop/recursive-async.rs
@@ -1,0 +1,11 @@
+//! Ensure DataflowConstProp doesn't cause an error with async recursion as in #155376.
+
+//@ edition:2018
+//@ check-pass
+//@ compile-flags: -Zmir-opt-level=0 -Zmir-enable-passes=+DataflowConstProp --crate-type=lib
+
+pub async fn foo(n: usize) {
+    if n > 0 {
+        Box::pin(foo(n - 1)).await;
+    }
+}

